tmk_core/avr.mk

```diff
@@ -2,7 +2,12 @@
 ##############################################################################
 # Compiler settings
 #
-CC = avr-gcc
+ifeq (,$(wildcard /usr/local/Cellar/avr-gcc\@8/*/bin/avr-gcc))
+    CC = /usr/local/Cellar/avr-gcc\@8/*/bin/avr-gcc
+else
+    CC = avr-gcc
+endif
```

Comment:
Travis does NOT like this.

Reply:
I don't like it either...

Reply:
Yeah me neither, its kinda bad practice to do this sort of thing.

Reply:
The problem I want to solve is that it's bad practice to force people to `brew link` gcc 8. Some people might want gcc 9 as their default because they work on things besides QMK. It leads to problems with the build environment that have to be sorted out, as the FAQ entries that would no longer be required show. 

I'm open to doing this other ways, but I think we can and should be able to use gcc8 over gcc9 when both are available.

Reply:
Most of my experience in the issue you mention is solved by adding the compiler you care about to be more important on the path. Maybe something that `qmk compile` could do instead?

Also the only 9 show stopper issue i'm aware of right now is 9.4 arm-none-eabi-gcc on arch.

Reply:
> Most of my experience in the issue you mention is solved by adding the compiler you care about to be more important on the path.

I've seen both ways done. There are no good answers in this situation.

The downside of PATH adjustments is that you can end up with a mixed toolchain situation. If a particular command, say avr-ld, gets removed it may find another version somewhere else in your PATH. This may work, or it may result in hard to find bugs.

The downside of absolute paths is more boilerplate code and detection, but in the event of a partially installed toolchain the problem is clear- the toolchain is missing a vital piece.

Reply:
Homebrew is not guaranteed to be installed to /usr/local. It's probably rare, but I imagine it does happen.
https://github.com/Homebrew/brew/blob/master/docs/Installation.md#untar-anywhere

We should at least be using `brew --cellar` to get the Cellar path.

Reply:
PATH isnt the only way to remove the need for OS specific absolute paths within the makefile.

`CC=/usr/bin/avr-gcc make -e lets_split:default` to which you can avoid the `-e` part if you change `avr.mk` to `CC ?= avr-gcc`.

Reply:
> We should at least be using brew --prefix <formula>.

I was not aware of that. Agreed that it should be used.

> PATH isnt the only way to remove the need for OS specific absolute paths within the makefile.

Are we going to force users to remember to do that every time they run make?

Reply:
> Are we going to force users to remember to do that every time they run make

Its fairly standard when dealing with cross compilers though I know that's not everyone in our user base. I was imagining something cli would deal with it, given our "makefiles are too complex, lets move the logic to something better" direction.

Reply:
I agree that CLI is the better place for it, but for right now we have a lot of users who are still compiling with `make keyboard:keymap`. I'd hate to create the situation where one of make or the CLI work and the other doesn't.

Reply:
> The downside of PATH adjustments is that you can end up with a mixed toolchain situation. If a particular command, say avr-ld, gets removed it may find another version somewhere else in your PATH. This may work, or it may result in hard to find bugs.

How is this solved by just setting `$CC`? Aren't you arguably more at risk of ending up with a mixed-toolchain situation by explicitly setting it? I can see you've got overrides for `$AR` and the like for ARM, but not AVR -- I'd imagine this is the sort of thing you're talking about. One other aspect which I've just noticed is that there's no setting for `$CXX` (even in master, now, for either AVR or ARM) -- and given that Lufa is half C++ it piques my curiosity as to why it works in the first place.

The only real way I've seen it consistently work is to specifically set `$PATH` -- something I already do in my QMK "wrapper" Makefile to point to the official ARM toolchain -- because I can't trust that the base OS environment is correct.

Either way, this change as it currently stands would've prevented me from testing the different versions of GCC when doing the ChibiOS upgrade without modifying the makefiles to roll this change back, so my vote here would be "no" on the explicit paths.

I don't, however, have a solution.

Reply:
> How is this solved by just setting $CC? Aren't you arguably more at risk of ending up with a mixed-toolchain situation by explicitly setting it?

If we make the check against a single binary (probably avr-gcc) and then set the path for the other binaries to the same directory it will fail with a "command not found" error should one of them not exist.

> I can see you've got overrides for $AR and the like for ARM, but not AVR 

When I unlinked avr-gcc those were still in /usr/local/bin. It seems that they're part of the `avr-bin-utils` package. I haven't had a chance to investigate to see if multiple versions of that package are installed when you have gcc8 and gcc9 installed at the same time.

> One other aspect which I've just noticed is that there's no setting for $CXX (even in master, now, for either AVR or ARM) -- and given that Lufa is half C++ it piques my curiosity as to why it works in the first place.

It's working here, I haven't looked into how yet.

> Either way, this change as it currently stands would've prevented me from testing the different versions of GCC when doing the ChibiOS upgrade without modifying the makefiles to roll this change back

If we use `:=` here so you could use environment vars to override would that be workable? I think this is something we should be doing even if we don't make any other changes to these definitions.

Reply:
> given that Lufa is half C++ it piques my curiosity as to why it works in the first place.

Most of the C++ code is in the usb_hid stuff, with a handful in ChibiOS, ugfx and QMK tests. I'm only seeing a single .cpp file in LUFA, and it's under BuildTests.

> I haven't had a chance to investigate to see if multiple versions of that package are installed when you have gcc8 and gcc9 installed at the same time.

I wouldn't imagine so; binutils is kinda independent of gcc.
https://github.com/osx-cross/homebrew-avr/tree/master/Formula only has one formula for avr-binutils, but several for the various versions of avr-gcc. They all depend on the same avr-binutils formula.